sdk/ml/azure-ai-ml/azure/ai/ml/_azure_environments.py

```diff
@@ -189,3 +198,70 @@ def _resource_to_scopes(resource):
     """
     scope = resource + "/.default"
     return [scope]
+
+def _get_registry_discovery_url(cloud, cloud_suffix):
+    """Get the registry discovery url
+        :return: string of discovery url
+    """
+    cloud_name = cloud["name"]
+    if cloud_name in _environments:
+        return _environments[cloud_name].registry_url
+    if ArmConstants.REGISTRY_ENV_URL in os.environ:
+        return os.environ[ArmConstants.REGISTRY_ENV_URL]
+    registry_discovery_region = os.environ.get(
+        ArmConstants.REGISTRY_DISCOVERY_REGION_ENV_NAME,
+        ArmConstants.REGISTRY_DISCOVERY_DEFAULT_REGION
+    )
+    return "https://{}{}.api.azureml.{}/".format(
+        cloud_name.lower(),
+        registry_discovery_region,
+        cloud_suffix
+    )
+
+def _get_clouds_by_metadata_url(metadata_url, timeout=ArmConstants.DEFAULT_TIMEOUT):
+    """Get all the clouds by the specified metadata url
+
+        :return: list of the clouds
+    """
+    import requests
+
+    try:
+        module_logger.debug('Start : Loading cloud metadata from the url specified by %s', metadata_url)
+        with requests.get(metadata_url, timeout=timeout) as meta_response:
```

Comment:
HttpPipeline is preferred over requests.

Reply:
Hi @paulshealy1, looking through how HttpPipeline works it seems that it's a universal object that needs to be passed in from the ml_client. Since this doesn't have a constructor, what would you recommend the refactor here be?
1. Change how the class works so that it has a constructor and everywhere calling it has to have an instance
2. Make the function calls require the pipeline as an argument, changing a bunch of places where it's called

I don't think I can just set the pipeline as a singleton in ml_client. I'm just wondering what the correct approach would be here, since this seems like it'll take a decent amount of code change to achieve.

Reply:
@singankit walked me through a solution here, disregard, thanks!